### DISCUSSION TITLE
How to run fuzzer in this project?

### DISCUSSION THREAD

**Discussion by qarmin** [slint-ui/slint#6495] [General]
I wanted to fuzz a little this project.

I see that already there was attempt to setup fuzzer, but PR was closed - https://github.com/slint-ui/slint/pull/1040/files

What functions should be tested to find the most crashes?

**Comment by ogoffart** [comment]
That PR was trying to fuzz the Slint parser.
But everything could be fuzzed on different level.
(rest of the compiler, software renderer, other renderer, keyboard/mouse input, ...)
This question is a bit abstract.

**Reply by qarmin** [reply]
Well, yes it is a bit abstract, but just a bit.

I don't really know internals of slint, so I don't know which functions can be used.
The easiest way is to test functions that as arguments takes random bytes/strings and produce result without using any global state, external libraries like x11/wayland or hardware devices like mouse/keyboard/speaker.

I can see that lexer and token parsing functions were used in u but they are relatively simple, so bugs would be hard to find.
Compilation is different and is much more advanced, so I think functions of this type would be good for testing, but I'm not sure if they exist

**Comment by hunger** [comment]
I never merged the fuzzer even though it did find a couple of issues.

The problem is that we recurse down the AST and that will eventually exhaust the stack... and the fuzzer started to get so reliable that it kept finding the recursion limits all the time.
